### PR TITLE
Pal 543 create k8s job to run example

### PR DESCRIPTION
**Pull request by developer01189998819991197253**

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -61,7 +61,59 @@ spec:
...
+      - name: docker-graph-storage
+        mountPath: /var/lib/docker
+    resources:
+      requests:
+        ephemeral-storage: "1Gi"
+      limits:
+        ephemeral-storage: "2Gi"
+
+  - name: maven
+    image: 779921734503.dkr.ecr.eu-west-1.amazonaws.com/jnlp-dood-new-infra:200608
```
Comment (ac74475) [inline Jenkinsfile:97]: So it is the same as the services one and using the newer helm version
```suggestion
    image: 779921734503.dkr.ecr.eu-west-1.amazonaws.com/jnlp-dood-new-infra:200710
```

File: deployment/local-k8s/k8s-bash-scripts/runFormattedK8sExample.sh [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
```
Comment (ac74475) [inline deployment/local-k8s/k8s-bash-scripts/runFormattedK8sExample.sh:1]: add licence header

File: deployment/local-k8s/k8s-bash-scripts/runK8sExample.sh [hunk 3]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
```
Comment (ac74475) [inline deployment/local-k8s/k8s-bash-scripts/runK8sExample.sh:1]: licence header

File: deployment/local-k8s/k8s-bash-scripts/verify.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
```
Comment (ac74475) [inline deployment/local-k8s/k8s-bash-scripts/verify.sh:1]: licence header

File: deployment/local-k8s/k8s-bash-scripts/verify.sh [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+FILE=./exampleOutput.txt
+
+if [ -f $FILE ]; then
+  num=$(wc -l $FILE | awk '{ print $1 }')
+  expected=764
+
+  # check the length of the exampleOutput text file and pass if it's 764
+  if [ $num == $expected ]; then
+    echo "Success - Number of lines was 764"
```
Comment (ac74475) [inline deployment/local-k8s/k8s-bash-scripts/verify.sh:11]:
```suggestion
    echo "Success - Number of lines was $expected"
```

File: deployment/local-k8s/k8s-bash-scripts/verify.sh [hunk 6]
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/env bash
+
+FILE=./exampleOutput.txt
+
+if [ -f $FILE ]; then
+  num=$(wc -l $FILE | awk '{ print $1 }')
+  expected=764
+
+  # check the length of the exampleOutput text file and pass if it's 764
```
Comment (ac74475) [inline deployment/local-k8s/k8s-bash-scripts/verify.sh:9]:
```suggestion
  # check the length of the exampleOutput text file and pass if it's the expected number of lines
```

File: deployment/local-k8s/local-bash-scripts/deployServicesToK8s.sh [hunk 7]
```diff
@@ -0,0 +1,6 @@
+#! /usr/bin/env bash
```
Comment (ac74475) [inline deployment/local-k8s/local-bash-scripts/deployServicesToK8s.sh:1]: licence header

File: deployment/local-k8s/local-bash-scripts/runFormattedK8sExample.sh [hunk 8]
```diff
@@ -0,0 +1,3 @@
+#! /usr/bin/env bash
```
Comment (ac74475) [inline deployment/local-k8s/local-bash-scripts/runFormattedK8sExample.sh:1]: licence header

File: deployment/local-k8s/local-bash-scripts/runK8sExample.sh [hunk 9]
```diff
@@ -0,0 +1,3 @@
+#! /usr/bin/env bash
```
Comment (ac74475) [inline deployment/local-k8s/local-bash-scripts/runK8sExample.sh:1]: licence header

File: deployment/local-k8s/local-bash-scripts/verify.sh [hunk 10]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
```
Comment (ac74475) [inline deployment/local-k8s/local-bash-scripts/verify.sh:1]: licence header

File: example-model/pom.xml [hunk 11]
```diff
@@ -90,11 +90,34 @@
                 <configuration>
                     <dateFormat>yyyyMMdd-HHmmss</dateFormat>
                     <dotGitDirectory>${project.basedir}/.git</dotGitDirectory>
-                    <generateGitPropertiesFile>false
-                    </generateGitPropertiesFile><!-- somehow necessary. otherwise the variables are not available in the pom -->
+                    <generateGitPropertiesFile>false</generateGitPropertiesFile><!-- somehow necessary. otherwise the variables are not available in the pom -->
                     <useNativeGit>true</useNativeGit>
                 </configuration>
             </plugin>
```
Comment (ac74475) [inline example-model/pom.xml:96]: Add pi (push image) profile

File: example-model/pom.xml [hunk 12]
```diff
@@ -90,11 +90,34 @@
...
+            <plugin>
```
Comment (ac74475) [inline example-model/pom.xml:97]: I would expect this build section to be the same as the other Palisade services

File: example-model/src/main/java/uk/gov/gchq/palisade/example/model/client/ExampleSimpleClient.java [hunk 13]
```diff
@@ -45,6 +45,11 @@ public void run(final String filename, final String userId, final String purpose
 
     public Stream<Employee> read(final String resourceId, final String userId, final String purpose) throws IOException {
         String absoluteFile = new File(resourceId).toURI().toString();
+        // Maintain is-a-file-or-a-directory consistency between relative and absolute paths
```
Comment (ac74475) [inline example-model/src/main/java/uk/gov/gchq/palisade/example/model/client/ExampleSimpleClient.java:48]: this comment doesn't really tell me what this is doing

File: example-model/src/main/resources/values.yaml [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+replicaCount: 1
+
+service:
+  port: 80
+  type: ClusterIP
+
+image:
+  name: @artifactId@
+  tag: @git.commit.id.abbrev@-@revision@
+  pullPolicy: IfNotPresent
```
Comment (ac74475) [inline example-model/src/main/resources/values.yaml:24]:
```suggestion
  pullPolicy: IfNotPresent
  codeRelease: @project.parent.version@
```

File: licenses/mit_license.html [hunk 15]
```diff
@@ -1,372 +0,0 @@
-<!DOCTYPE html>
```
Comment (ac74475) [inline licenses/mit_license.html:1]: add licence back in

File: values.yaml [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+#
+
+# Use Palisade-services values.yaml mostly, but set some useful defaults here
+
+# Used by templates for naming resources, see assorted helpers.tpl
+fullnameOverride: palisade-examples
+
+global:
+  # uniqueNamespace: if true (and this is a new install) the generated deployment name will be used as the namespace name too
+  uniqueNamespace: false
```
Comment (ac74475) [inline values.yaml:23]: Am not sure on this mechanism for namespacing, can we just have the user supply a --namespace my-namespace when they do the helm update and it work out if you need a new namespace or not. And we can default to palisade-examples as the namespace if none is provided.

File: code-style/jetcd-suppression.xml [hunk 17]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0"?>
+
+<!DOCTYPE suppressions PUBLIC
+        "-//Puppy Crawl//DTD Suppressions 1.1//EN"
+        "http://www.puppycrawl.com/dtds/suppressions_1_1.dtd">
+
+<suppressions>
+    <suppress checks="." files="io[\\/]etcd[\\/]jetcd[\\/]launcher"/>
+</suppressions>
```
[the remark this answers is not in the document]
Comment (developer01189998819991197253) [reply code-style/jetcd-suppression.xml:9]: removed

File: deployment/bash-scripts/setScriptPath.sh [hunk 18]
```diff
@@ -1,7 +0,0 @@
-#!/bin/bash
```
[the remark this answers is not in the document]
Comment (developer01189998819991197253) [reply deployment/bash-scripts/setScriptPath.sh:1]: readded to original location

File: charts/example-model/templates/deployment.yaml [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+    matchLabels:
+      app.kubernetes.io/name: {{ include "example-model.name" . }}
+      app.kubernetes.io/instance: {{ .Release.Name }}
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: {{ include "example-model.name" . }}
+        app.kubernetes.io/instance: {{ .Release.Name }}
+    spec:
+      serviceAccountName: internal-kubectl
```
[the remark this answers is not in the document]
Comment (developer01189998819991197253) [reply charts/example-model/templates/deployment.yaml:37]: renamed

File: deployment/local-jvm/bash-scripts/verify.sh [hunk 20]
```diff
@@ -1,47 +1,44 @@
...
-
   if [ -f $FILE ]; then
     num=$(wc -l $FILE | awk '{ print $1 }')
+    expected=764
 
-    #check the length of the exampleOutput text file and pass if it's 764
-    if [ $num == 764 ]; then
+    # check the length of the exampleOutput text file and pass if it's 764
+    if [ $num == $expected ]; then
       echo "Success - Number of lines was 764"
```
[the remark this answers is not in the document]
Comment (developer01189998819991197253) [reply deployment/local-jvm/bash-scripts/verify.sh:13]: added to local commit

File: deployment/local-jvm/bash-scripts/verify.sh [hunk 21]
```diff
@@ -1,47 +1,44 @@
...
-  # Important to cd before running the jar - the working directory must be deployment/local-jvm/bash-scripts/
-  cd $DIR
-
   if [ -f $FILE ]; then
     num=$(wc -l $FILE | awk '{ print $1 }')
+    expected=764
 
-    #check the length of the exampleOutput text file and pass if it's 764
-    if [ $num == 764 ]; then
+    # check the length of the exampleOutput text file and pass if it's 764
```
[the remark this answers is not in the document]
Comment (developer01189998819991197253) [reply deployment/local-jvm/bash-scripts/verify.sh:11]: added to local commit

File: deployment/local-k8s/k8s-bash-scripts/runFormattedK8sExample.sh (hunk 2 above)
Comment (developer01189998819991197253) [reply deployment/local-k8s/k8s-bash-scripts/runFormattedK8sExample.sh:1]: added

File: deployment/local-k8s/k8s-bash-scripts/verify.sh (hunk 4 above)
Comment (developer01189998819991197253) [reply deployment/local-k8s/k8s-bash-scripts/verify.sh:1]: added

File: deployment/local-k8s/k8s-bash-scripts/runK8sExample.sh (hunk 3 above)
Comment (developer01189998819991197253) [reply deployment/local-k8s/k8s-bash-scripts/runK8sExample.sh:1]: added

File: deployment/local-k8s/k8s-bash-scripts/verify.sh (hunk 5 above)
Comment (developer01189998819991197253) [reply deployment/local-k8s/k8s-bash-scripts/verify.sh:11]: added to local commit

File: deployment/local-k8s/k8s-bash-scripts/verify.sh (hunk 6 above)
Comment (developer01189998819991197253) [reply deployment/local-k8s/k8s-bash-scripts/verify.sh:9]: added to local commit

File: deployment/local-k8s/local-bash-scripts/deployServicesToK8s.sh (hunk 7 above)
Comment (developer01189998819991197253) [reply deployment/local-k8s/local-bash-scripts/deployServicesToK8s.sh:1]: added

File: deployment/local-k8s/local-bash-scripts/runFormattedK8sExample.sh (hunk 8 above)
Comment (developer01189998819991197253) [reply deployment/local-k8s/local-bash-scripts/runFormattedK8sExample.sh:1]: added

File: deployment/local-k8s/local-bash-scripts/runK8sExample.sh (hunk 9 above)
Comment (developer01189998819991197253) [reply deployment/local-k8s/local-bash-scripts/runK8sExample.sh:1]: added

File: deployment/local-k8s/local-bash-scripts/verify.sh (hunk 10 above)
Comment (developer01189998819991197253) [reply deployment/local-k8s/local-bash-scripts/verify.sh:1]: added

File: example-model/pom.xml (hunk 11 above)
Comment (developer01189998819991197253) [reply example-model/pom.xml:96]: added

File: licenses/mit_license.html (hunk 15 above)
Comment (developer01189998819991197253) [reply licenses/mit_license.html:1]: readded

File: charts/example-model/templates/service-account.yaml [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: internal-kubectl
```
[the remark this answers is not in the document]
Comment (developer01189998819991197253) [reply charts/example-model/templates/service-account.yaml:18]: added

File: charts/example-model/templates/role.yaml [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: get-pods-logs
```
[the remark this answers is not in the document]
Comment (developer01189998819991197253) [reply charts/example-model/templates/role.yaml:18]: added

File: charts/example-model/templates/role-binding.yaml [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: get-pods-logs-to-sa
```
[the remark this answers is not in the document]
Comment (developer01189998819991197253) [reply charts/example-model/templates/role-binding.yaml:18]: added

File: example-model/src/main/resources/application-k8s.yaml [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+web:
+  client:
+    palisade-service: "http://palisade-service"
+    data-service: "http://data-service"
```
Comment (dev930018) [inline example-model/src/main/resources/application-k8s.yaml:19]: http:// shouldn't be needed here, as discussed elsewhere
data-service: ... shouldn't be needed here, it is picked up from the connection detail

Comment (developer01189998819991197253) [reply example-model/src/main/resources/application-k8s.yaml:19]: removed http
removed data service

File: example-model/pom.xml [hunk 26]
```diff
@@ -150,6 +173,62 @@
...
+                            </files>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-resources-plugin</artifactId>
+                <version>3.1.0</version>
+            </plugin>
```
Comment (dev930018) [inline example-model/pom.xml:198]: Redundant, see other maven-resources-plugin above

**Comment by dev930018** [comment]
Looking at k8s performance tests, it might be better to move the deployment/local-k8s/k8s-bash-scripts to the example-model resources, then deployment/local-k8s contains the contents of (the currently nested) local-bash-scripts

This would make sense when it comes to k8s performance tests having another set of k8s bash scripts

File: example-model/pom.xml (hunk 26 above)
Comment (developer01189998819991197253) [reply example-model/pom.xml]: removed

File: example-model/src/main/java/uk/gov/gchq/palisade/example/model/client/ExampleSimpleClient.java (hunk 13 above)
Comment (developer01189998819991197253) [reply example-model/src/main/java/uk/gov/gchq/palisade/example/model/client/ExampleSimpleClient.java]: updated comment

File: example-model/pom.xml (hunk 12 above)
Comment (developer01189998819991197253) [reply example-model/pom.xml:141]: updated build section and plugins, ran mvn install -P pi and tags updated

File: values.yaml (hunk 16 above)
Comment (developer01189998819991197253) [reply values.yaml:23]: updated

File: deployment/hr-data-generator/createHRData.sh [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+then
+    JAR_FILE=$(find "$TARGET_DIR" -type f -iname "hr-data-generator-*-jar-with-dependencies.jar")
+    if [ ! -z "$JAR_FILE" ];
+    then
+        FILE_PRESENT=1
+    fi
+fi
+
+if [ "$FILE_PRESENT" -eq 0 ];then
+    echo "Can't find hr-data-generator-<version>.jar in ${TARGET_DIR}. Have you run \"mvn install -P example\" ?"
```
Comment (ac74475) [inline deployment/hr-data-generator/createHRData.sh:31]:
```suggestion
    echo "Can't find hr-data-generator-<version>.jar in ${TARGET_DIR}. Have you run \"mvn install\" ?"
```

**Pull request merged**